### DISCUSSION TITLE
Create multichannel file (psd, tif) from single tif files

### DISCUSSION THREAD

**Discussion by tresorjoe** [ImageMagick/ImageMagick#6833] [Help]
### ImageMagick version

7.1.0

### Operating system, version and so on

macOS Monterey 12.7.1

### Description

Is there a way I can merge multiple single tif files (e.g. 8) into a multichannel psd/tif file?
I've read that it's basically only possible to create RGBA or CMYKA files but looking at the multispectral imagery topic it got me curious. Thanks.

**Comment by snibgo** [comment]
Suppose you have two input tiffs, each containing just RGB channels. You want an output tiff that has the RGB channels from the first input, and three meta channels from the RGB channels of the second input.
```
magick x1.tiff x2.tiff -channel-fx "| red=>meta0 green=>meta1 blue=>meta2" x4.tiff
```
We can verify the output `x4.tiff` with `exiftool` or IM's `-verbose info:`", or `ftxt:` like this:
```
magick x4.tiff -define ftxt:format="\p\n" ftxt:
```
If you have 8 input tiffs, you need a longer `-channel-fx`:
```
magick x1.tiff x2.tiff x3.tiff x4.tiff x5.tiff x6.tiff x7.tiff x8.tiff -channel-fx "| red=>meta0 green=>meta1 blue=>meta2 | red=>meta3 green=>meta4 blue=>meta5 | red=>meta6 green=>meta7 blue=>meta8 | red=>meta9 green=>meta10 blue=>meta11 | red=>meta12 green=>meta13 blue=>meta14 | red=>meta15 green=>meta16 blue=>meta17 | red=>meta18 green=>meta19 blue=>meta20" xout.tiff
```

**Reply by tresorjoe** [reply]
Thank you @snibgo 
I forgot to mention that the source files are 8 or 16bit grayscale images (not RGB). In some cases they may also be 1bit or 2bit but that shouldn't be an issue since I know how to bring them to 8bit. Would the procedure be identical as above or would I have to treat grayscale source images a bit differently?

**Comment by snibgo** [comment]
If the inputs are 1 bit per channel, or 8 or 16 or 32, floating-point or integer, makes no difference to the command.

If the inputs are only one channel, that does make a difference, because you will need to create just one meta channel per image. However, there seems to be a bug in IM v7.1.1-20.

I create four TIFFs for example inputs.
```
magick xc:gray(10%%) x1.tiff

magick xc:gray(22%%) x2.tiff

magick xc:gray(33%%) x3.tiff

magick xc:gray(44%%) x4.tiff
```
The next command should make an output with one gray channel, and three meta channels, and these should be populated from the four inputs.
```
magick x1.tiff x2.tiff x3.tiff x4.tiff -channel-fx "| gray=>meta0 | gray=>meta1 | gray=>meta2" g4.tiff
```
`-verbose info` shows that `g4.tiff` contains RGB channels (which seems wrong) and three meta channels. The RGB channes are given values from x1.tiff, x2.tiff and x3.tiff respectively. The first meta channel (`meta0`) is given the value from x4.tiff. The other two meta channels are zero.

This seems to be a bug.

**Reply by tresorjoe** [reply]
Thank you. I can reproduce. Would you keep me posted on a possible bugfix?

**Comment by snibgo** [comment]
A developer may respond to this thread.

**Comment by fmw42** [comment]
Workaround.  Use -combine on the first 3 images, then add the 4th as meta.

`magick x1.tiff x2.tiff x3.tiff -set colorspace sRGB -combine x4.tiff -channel-fx "| gray=>meta0 g4.tiff`

*Correct this if I have the syntax in error.*

**Reply by tresorjoe** [reply]
Thank you @fmw42 
Unfortunately this solution creates an RGB + channels file as far as I understand. I would need a native multichannel file.

**Reply by fmw42** [reply]
Sorry, I misunderstood what you wanted.

**Comment by snibgo** [comment]
We should be able to do this via `ftxt:` format (assuming the inputs have the same dimensions). First, write the input images to text files:
```
magick x1.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t1.txt
magick x2.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t2.txt
magick x3.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t3.txt
magick x4.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t4.txt
```
Append corresponding text line together:
```
paste --delimiter=, t1.txt t2.txt t3.txt t4.txt >out.txt
```
(`paste` is a common Unix utility.)

My examples have only one pixel. `out.txt` looks like this:
```
10.0007629510948%,22.0004577706569%,33.0006866559854%,43.9993896391241%
```
Read that file as an image size 1x1 that is grayscale with 3 meta channels:
```
magick -size 1x1 -colorspace Gray -define ftxt:nummeta=3 -define ftxt:format="\p\n" ftxt:out.txt -verbose +write info: out.tiff
```
The `-verbose +write info:` shows the image has, as required, one gray channel and 3 meta channels. Sadly, `out.tiff` contains (as before) RGB channels and 3 meta channels, and the numbers have been put into the 3 RGB channels and just the first meta channel.

This seems to be a bug when IM writes a grayscale image with meta channels to a tiff file.

**Reply by urban-warrior** [reply]
Give us a few days to review the issue and report back.

**Comment by snibgo** [comment]
A simpler example shows the problem:
```
magick xc:gray(10%) -channel-fx "gray=>meta0" -verbose info:
```
The `info:` shows correctly that we have a gray channel and a meta0 channel, and they are both 10%. This is correct.
```
magick xc:gray(10%) -channel-fx "gray=>meta0" x.tiff

magick x.tiff -verbose info:
```
The `info:` shows that x.tiff has RGB channels and meta0. The Red and Green channels are 10%, the Blue and meta0 channels are zero. This is wrong.

**Comment by snibgo** [comment]
I think the problem is in tiff.c function WriteTIFFImage(), in lines like:
```
                (void) TIFFSetField(tiff,TIFFTAG_SAMPLESPERPIXEL,1);

```
Currently, the TIFFTAG_SAMPLESPERPIXEL field is set to `1` only for grayscale images and `image->number_meta_channels == 0`.

When `extra_samples != 0`, this is called:
```
        (void) TIFFGetFieldDefaulted(tiff,TIFFTAG_SAMPLESPERPIXEL,
          &samples_per_pixel,sans);
```
... and that sets samples_per_pixel to 3, because the earlier code that tests for grayscale isn't called because there are meta channels.

In the next line:
```
        (void) TIFFSetField(tiff,TIFFTAG_SAMPLESPERPIXEL,samples_per_pixel+
          extra_samples);
```
.. so the TIFFTAG_SAMPLESPERPIXEL field is set to 3 plus the number of metachannels, even for grayscale images.

So the patch is:

```
--- "coders\\tiff231101.c"	2023-11-01 14:33:37.672832600 +0000
+++ "coders\\tiff.c"	2023-11-01 14:43:32.269581900 +0000
@@ -3676,8 +3676,7 @@
             photometric=PHOTOMETRIC_RGB;
         (void) TIFFSetField(tiff,TIFFTAG_SAMPLESPERPIXEL,3);
         if ((image_info->type != TrueColorType) &&
-            (image_info->type != TrueColorAlphaType) &&
-            (image->number_meta_channels == 0))
+            (image_info->type != TrueColorAlphaType))
           {
             if ((image_info->type != PaletteType) &&
                 (IdentifyImageCoderGray(image,exception) != MagickFalse))
```
I haven't tested this extensively.

**Reply by urban-warrior** [reply]
Patch applied.

**Comment by snibgo** [comment]
With the above patch to `tiff.c`, the OP problem can be solved using `ftxt:` as shown above:
```
magick xc:gray(10%%) x1.tiff
magick xc:gray(22%%) x2.tiff
magick xc:gray(33%%) x3.tiff
magick xc:gray(44%%) x4.tiff

magick x1.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t1.txt
magick x2.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t2.txt
magick x3.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t3.txt
magick x4.tiff -precision 15 -define ftxt:format="\p\n" ftxt:t4.txt

paste --delimiter=, t1.txt t2.txt t3.txt t4.txt >out.txt

magick -size 1x1 -colorspace Gray -define ftxt:nummeta=3 -define ftxt:format="\p\n" ftxt:out.txt -verbose +write info: out.tiff
```
`out.tiff` then contains a grayscale channel and 3 meta channels, and the four values are 10%, 22%, 33% and 44%, as required.

Sadly, the solution with `-channel-fx` doesn't work correctly.
```
magick x1.tiff x2.tiff x3.tiff x4.tiff -colorspace gray -channel-fx "| gray=>meta0 | gray=>meta1 | gray=>meta2" -verbose +write info: g4.tiff
```
The `+write info:` and `g4.tiff` contain the correct channels (one grayscale and three meta), but the channels contain the wrong values: 10%, 22%, 22% and 22%.

So this appears to be a bug in `-channel-fx` (or in my understanding of `-channel-fx`). I'll look at that.

**Comment by snibgo** [comment]
On `-channel-fx`, in `channel.c` the comments about `ChannelFxImage()` claim:
```
%    |       add a new output image for the next set of channel operations
%    ;       move to the next input image for the source of channel data
```
These are the wrong way round.

In the code, the `|` symbol actually means:

- If the current image list contains only one image, move to that image for the source of channel data.

- If the current image list contains exactly two images, move to the second image for the source of channel data.

- If the current image list more than two images, move to the second image for the source of channel data.

So it is not possible to move to the third or fourth etc input images.

I think a better meaning for `|` is: "move to the next input image for the source of channel data. If there are no more images in the list, | has no effect."

Here is the patch for channel.c:
```
--- "magickcore\\channel231101.c"	2023-11-01 17:52:20.012878300 +0000
+++ "magickcore\\channel.c"	2023-11-01 18:02:11.403498200 +0000
@@ -79,8 +79,9 @@
 %    =>      copy one channel to another channel (e.g. red=>green)
 %    =       assign a constant value to a channel (e.g. red=50%)
 %    ,       write new image channels in the specified order (e.g. red, green)
-%    |       add a new output image for the next set of channel operations
-%    ;       move to the next input image for the source of channel data
+%    ;       add a new output image for the next set of channel operations
+%    |       move to the next input image for the source of channel data
+%             If there are no more images in the list, | has no effect.
 %
 %  For example, to create 3 grayscale images from the red, green, and blue
 %  channels of an image, use:
@@ -269,7 +270,6 @@
       }
       case '|':
       {
-        source_image=GetFirstImageInList(source_image);
         if (GetNextImageInList(source_image) != (Image *) NULL)
           source_image=GetNextImageInList(source_image);
         (void) GetNextToken(p,&p,MagickPathExtent,token);
```
With this patch, and the tiff.c patch above, the following command...
```
magick x1.tiff x2.tiff x3.tiff x4.tiff -colorspace gray -channel-fx "| gray=>meta0 | gray=>meta1 | gray=>meta2" -verbose +write info: g4.tiff
```
... works as expected.

[Edited to remove formatting error.]

**Reply by urban-warrior** [reply]
Thanks for the analysis and patch.  Patch applied.

**Comment by tresorjoe** [comment]
Hi @snibgo and @urban-warrior 
many thanks for your analysis and bugfixing. Sounds promising. Since I am not a developer myself I was wondering whether/how I can test it with e.g. pre-compiled binaries for macOS. Is a new version scheduled soon? Or do I just "brew install imagemagick" and it will update automatically incl. your last fixes?

**Comment by snibgo** [comment]
The patches have been applied to the source code, so any process that builds from source will include the patches. I don't know if "brew install imagemagick" builds from source.

Sorry, I don't know release dates for pre-complied binaries.
